### PR TITLE
Backport new PQ implementation to 5.6 

### PR DESCRIPTION
**Pull request by original-brownbear**
Exact back port/copy of Java portion of the PQ code with the exception of the version byte still being `1` here and no version file being written (the upgrade logic isn't in here).

### CODE REVIEW

**Comment by original-brownbear** [comment]
@andrewvc PHEW :D ... but seems to work fine as far as I can tell:) Will test a little more soon.

**Comment by original-brownbear** [comment]
CC: @gmoskovicz @tsg

File: logstash-core/src/main/java/org/logstash/ackedqueue/HeadPage.java [hunk 1]
```diff
@@ -1,127 +0,0 @@
-package org.logstash.ackedqueue;
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/ackedqueue/HeadPage.java]: Confirmed as deleted in master

File: logstash-core/src/main/java/org/logstash/ackedqueue/QueueRuntimeException.java [hunk 2]
```diff
@@ -2,28 +2,8 @@
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/ackedqueue/QueueRuntimeException.java]: Confirmed no diff with master

File: logstash-core/src/main/java/org/logstash/ackedqueue/Settings.java [hunk 3]
```diff
@@ -1,14 +1,7 @@
 package org.logstash.ackedqueue;
```
Comment (andrewvc) [inline logstash-core/src/main/java/org/logstash/ackedqueue/Settings.java]: No diff with master

File: logstash-core/src/main/java/org/logstash/ackedqueue/Page.java [hunk 4]
```diff
@@ -1,28 +1,28 @@
 package org.logstash.ackedqueue;
```
[the remark this answers is not in the document]
Comment (original-brownbear) [reply logstash-core/src/main/java/org/logstash/ackedqueue/Page.java]: Jup, didn't wanna add the dependency for that one :)

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -57,7 +57,7 @@ def puts(payload)
     # Some tests mess with the settings. This ensures one test cannot pollute another
     LogStash::SETTINGS.reset
 
-    LogStash::SETTINGS.set("queue.type", "memory_acked")
+    LogStash::SETTINGS.set("queue.type", "persisted")
```
Comment (yaauie) [inline spec/spec_helper.rb]: On `master`, `6.x`, and `6.3`, we're currently using `memory` in specs for testing stability reasons (see: https://github.com/elastic/logstash/pull/9601); do we want to do that here too?

Comment (original-brownbear) [reply spec/spec_helper.rb]: @yaauie this was on purpose, trying to get a little extra coverage given the circumstances :)

**Comment by original-brownbear** [comment]
@yaauie good to merge here?

**Comment by original-brownbear** [comment]
@yaauie thanks merging then :)

**Comment by original-brownbear** [comment]
> If possible, adding a log message indicating that it is loading the persistent queue would greatly improve the experience -- once Logstash launched, it appeared stalled for approximately 2 minutes before it started emitting items.

Phew let's not do it here (no risks on this one please :D), but I agree we should look into some visual feedback there. The initial loading of the queue is a **really** messed up thing UI wise at the moment.

**Pull request merged**